File: sasu-backend/src/main/java/com/sasu/family/model/Expense.java
package com.sasu.family.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;

@Entity
@Table(name = "expenses")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Expense {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private BigDecimal amount;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ExpenseCategory category;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Frequency frequency;

    private LocalDate startDate;

    private String description;

    @Column(nullable = false)
    private Boolean active = true;

    @Column(nullable = false)
    private Boolean isNeed = true; // true for Needs, false for Wants

    public enum ExpenseCategory {
        FOOD,
        UTILITIES,
        TRANSPORTATION,
        EDUCATION,
        HEALTHCARE,
        ENTERTAINMENT,
        SHOPPING,
        HOUSING,
        INSURANCE,
        SAVINGS,
        OTHER
    }

    public enum Frequency {
        MONTHLY,
        QUARTERLY,
        YEARLY,
        ONE_TIME
    }

    public BigDecimal getMonthlyAmount() {
        if (frequency == Frequency.MONTHLY) {
            return amount;
        } else if (frequency == Frequency.QUARTERLY) {
            return amount.divide(BigDecimal.valueOf(3), 2, BigDecimal.ROUND_HALF_UP);
        } else if (frequency == Frequency.YEARLY) {
            return amount.divide(BigDecimal.valueOf(12), 2, BigDecimal.ROUND_HALF_UP);
        }
        return BigDecimal.ZERO;
    }
}

File: sasu-mobile/lib/data/models/expense.dart
/// Expense Model
///
/// Represents a family expense

class Expense {
  final int? id;
  final String name;
  final double amount;
  final String category;
  final String frequency;
  final String? startDate;
  final String? description;
  final bool active;
  final bool isNeed; // true for Needs, false for Wants

  Expense({
    this.id,
    required this.name,
    required this.amount,
    required this.category,
    required this.frequency,
    this.startDate,
    this.description,
    this.active = true,
    this.isNeed = true,
  });

  factory Expense.fromJson(Map<String, dynamic> json) {
    return Expense(
      id: json['id'],
      name: json['name'] ?? '',
      amount: (json['amount'] ?? 0).toDouble(),
      category: json['category'] ?? '',
      frequency: json['frequency'] ?? '',
      startDate: json['startDate'],
      description: json['description'],
      active: json['active'] ?? true,
      isNeed: json['isNeed'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'amount': amount,
      'category': category,
      'frequency': frequency,
      'startDate': startDate,
      'description': description,
      'active': active,
      'isNeed': isNeed,
    };
  }

  double getMonthlyAmount() {
    if (frequency == 'MONTHLY') {
      return amount;
    } else if (frequency == 'QUARTERLY') {
      return amount / 3;
    } else if (frequency == 'YEARLY') {
      return amount / 12;
    }
    return 0;
  }
}

